File: storage/native/src/tarantool_factory.c
#include "tarantool_factory.h"
#include <asm-generic/errno-base.h>
#include "interactor_message.h"
#include "memory_module.h"
#include "small.h"

int tarantool_factory_initialize(struct tarantool_factory* factory, struct tarantool_factory_configuration* configuration)
{
    float actual_alloc_factor;

    factory->memory = calloc(1, sizeof(struct memory));
    if (!factory->memory)
    {
        return -ENOMEM;
    }
    if (memory_create(factory->memory, configuration->quota_size, configuration->preallocation_size, configuration->slab_size))
    {
        return -ENOMEM;
    }

    factory->tarantool_datas = calloc(1, sizeof(struct small_alloc));
    if (!factory->tarantool_datas)
    {
        return -ENOMEM;
    }
    small_alloc_create(factory->tarantool_datas, &factory->memory->cache, 3 * sizeof(int), sizeof(intptr_t), 1.05, &actual_alloc_factor);

    return 0;
}

const char* tarantool_create_string(struct tarantool_factory* factory, size_t size)
{
    return smalloc(factory->tarantool_datas, size);
}

void tarantool_free_string(struct tarantool_factory* factory, const char* string, size_t size)
{
    smfree(factory->tarantool_datas, (void*)string, size);
}

void tarantool_factory_destroy(struct tarantool_factory* factory)
{
    small_alloc_destroy(factory->tarantool_datas);
    memory_destroy(factory->memory);
    free(factory->tarantool_datas);
    free(factory->memory);
}

File: storage/dart/lib/bindings/storage/tarantool_factory.dart
// Generated
// ignore_for_file: unused_import

import 'dart:ffi';
import 'package:ffi/ffi.dart';
import '../../storage/bindings.dart';

final class tarantool_factory extends Struct {
  external Pointer<memory_instance> memory;
  external Pointer<memory_small_allocator> tarantool_datas;
}

final class tarantool_factory_configuration extends Struct {
  @Size()
  external int quota_size;
  @Size()
  external int slab_size;
  @Size()
  external int preallocation_size;
}

@Native<Int32 Function(Pointer<tarantool_factory> factory, Pointer<tarantool_factory_configuration> configuration)>(isLeaf: true)
external int tarantool_factory_initialize(Pointer<tarantool_factory> factory, Pointer<tarantool_factory_configuration> configuration);

@Native<Pointer<Utf8> Function(Pointer<tarantool_factory> factory, Size size)>(isLeaf: true)
external Pointer<Utf8> tarantool_create_string(Pointer<tarantool_factory> factory, int size);

@Native<Void Function(Pointer<tarantool_factory> factory, Pointer<Utf8> string, Size size)>(isLeaf: true)
external void tarantool_free_string(Pointer<tarantool_factory> factory, Pointer<Utf8> string, int size);

@Native<Void Function(Pointer<tarantool_factory> factory)>(isLeaf: true)
external void tarantool_factory_destroy(Pointer<tarantool_factory> factory);
